packages/core/src/rules/common/component-name-unique.ts

```diff
@@ -0,0 +1,158 @@
+  [TYPE_NAME_RESPONSE]: 'response',
+  [TYPE_NAME_REQUEST_BODY]: 'requestBody',
+};
+
+export const ComponentNameUnique: Oas3Rule | Oas2Rule = (options) => {
+  const components = new Map<string, Set<string>>();
+
+  const typeNames: string[] = [];
+  if (options.schema !== 'off') {
+    typeNames.push(TYPE_NAME_SCHEMA);
+  }
+  if (options.parameter !== 'off') {
+    typeNames.push(TYPE_NAME_PARAMETER);
+  }
+  if (options.response !== 'off') {
+    typeNames.push(TYPE_NAME_RESPONSE);
+  }
+  if (options.requestBody !== 'off') {
+    typeNames.push(TYPE_NAME_REQUEST_BODY);
+  }
+
+  function getOptionNameForTypeName(typeName: string): string | null {
+    return TYPE_NAME_TO_OPTION_NAME[typeName] ?? null;
+  }
+
+  function getComponentNameFromAbsoluteLocation(absoluteLocation: string): string {
+    const componentName = absoluteLocation.split('/').slice(-1)[0];
+    if (
+      componentName.endsWith('.yml') ||
+      componentName.endsWith('.yaml') ||
+      componentName.endsWith('.json')
+    ) {
+      return componentName.slice(0, componentName.lastIndexOf('.'));
+    }
+    return componentName;
+  }
+
+  function addFoundComponent(
+    typeName: string,
+    componentName: string,
+    absoluteLocation: string
+  ): void {
+    const key = getKeyForComponent(typeName, componentName);
+    const locations = components.get(key) ?? new Set();
+    locations.add(absoluteLocation);
+    components.set(key, locations);
+  }
+
+  function getKeyForComponent(typeName: string, componentName: string): string {
+    return `${typeName}/${componentName}`;
+  }
+  function getComponentFromKey(key: string): { typeName: string; componentName: string } {
+    const [typeName, componentName] = key.split('/');
+    return { typeName, componentName };
+  }
+
+  function addComponentFromAbsoluteLocation(typeName: string, absoluteLocation: string): void {
+    const componentName = getComponentNameFromAbsoluteLocation(absoluteLocation);
+    addFoundComponent(typeName, componentName, absoluteLocation);
+  }
+
+  const rule: Oas3Visitor = {
+    ref: {
+      leave(ref: OasRef, { type, resolve }: UserContext) {
+        const typeName = type.name;
+        if (typeNames.includes(typeName)) {
+          const resolvedRef = resolve(ref);
+          if (!resolvedRef.location) return;
+
+          addComponentFromAbsoluteLocation(
+            typeName,
+            resolvedRef.location.absolutePointer.toString()
+          );
+        }
+      },
+    },
+    Root: {
+      leave(root: Oas3Definition | Oas2Definition, ctx: UserContext) {
+        components.forEach((value, key, _) => {
+          if (value.size > 1) {
+            const component = getComponentFromKey(key);
+            const definitions = Array.from(value)
+              .map((v) => `- ${v}`)
+              .join('\n');
+
+            const problem: Problem = {
+              message: `${component.typeName} '${component.componentName}' is not unique. It is defined at:\n${definitions}`,
```

Comment:
How do you feel about this message?

```suggestion
              message: `Component '${component.typeName}/${component.componentName}' is not unique. It is defined at:\n${definitions}`,
```

Reply:
Not 100% sure it is right with the `/`. As normally it is `parameters/ParameterOne` but with the code currently it would be `Parameter/ParameterOne`.

What do you think about `Parameter component 'ParameterOne' is no unique...` ? Have pushed it with this change for now.

The other way would be to use the mapping I have to convert `Parameter` into `parameters` and then combining with `/` will work. But it becomes again a bit more complex. WDYT?

Reply:
Actually went with `parameters/ParameterOne` now (pushed) 😄. The output is nicer for the user.